[Fix] `no-invalid-html-attribute`: allow 'shortcut icon' on `link`

Primajin opened the pull request:
Closes #3172 

## Description

For historical reasons, the icon keyword may be preceded by the keyword "shortcut".

[https://html.spec.whatwg.org/multipage/links.html#rel-icon](https://html.spec.whatwg.org/multipage/links.html#rel-icon) (last sentence before the next paragraph about `license`)

## Intentions

1. First of all I'm checking if there are pairs registered for the given attribute name after all.
2. Now I'm splitting the attribute value(s) into *overlapping* pairs. If you want to check on the funky regex: https://regex101.com/r/aJXAEz/1
3. now I'm iterating pair by pair
4. on each pair I check what are the possible siblings for my current item - currently there is only 1:1 but potentially there could be a pair where `foo` can go with `bar` OR `baz` but NEVER `qux`
5. now I check if my pair is separated by a single space or multiple in between
6. Check if the first value has a corresponding pairing
7. Get the last value as this is the "real" second one
8. Check if this is not one of the anticipated siblings
9. set message and message id based on, if the second value is truthy
    1. if it is the attribute was not paired with the right one
    2. if its not it stands alone which makes it invalid.
10. report the issue
11. check if there is only one option what the sibling could be, if there is only one anyways let's autofix it for the user! 🎉 

## Demo

This introduces two new messages:

```html
<link rel="shortcut"></link>
```
Reports: `“shortcut” must be directly followed by “icon”.`

<br><br>

```html
<link rel="shortcut manifest"></link>
```
Reports: `“shortcut” can not be directly followed by “manifest” without “icon”.`<br>

ljharb commented on tests/lib/rules/no-invalid-html-attribute.js:
```diff
@@ -130,6 +130,9 @@ ruleTester.run('no-invalid-html-attribute', rule, {
     { code: '<link rel="icon"></link>' },
     { code: 'React.createElement("link", { rel: "icon" })' },
     { code: 'React.createElement("link", { rel: ["icon"] })' },
+    { code: '<link rel="shortcut icon"></link>' },
+    { code: 'React.createElement("link", { rel: "shortcut icon" })' },
+    { code: 'React.createElement("link", { rel: ["shortcut icon"] })' },
```
these should be valid, but shouldn't `shortcut` and `shortcut foo` and `foo shortcut` all be invalid?

Primajin replied:
Yep but I am having a hard time wrapping my head around the code and the error messages.

So normally `neverValid` would trigger - which is kind of not true - because paired with `icon` it is valid on link.
I changed the Map item to `['shortcut\u0020icon', new Set(['link'])],` which also doesn't help because then it still reports `shortcut` to be never valid and I don't want to allow `shortcut` to be valid alone 🤔

ljharb replied:
since this is a special case of two words, we might need to alter the rule where it splits on a space.

Primajin replied:
Not feeling super comfortable doing that on my first contribution to this project though 😅  Wouldn't really know where to start and keep it clean.

ljharb replied:
That’s what code review is for, if you’re up for it :-)

Primajin replied:
Alright I'll take a stab at it 🔪

Primajin replied:
OK did some progress, need to take another look regarding multiple spaces, signing off for today 😄

ljharb commented on lib/rules/no-invalid-html-attribute.js:
```diff
@@ -51,152 +57,160 @@ const VALID_VALUES = new Map([
+  'area',
   'article',
   'aside',
-  'footer',
```
let's revert sorting this list so the diff is cleaner

Primajin replied:
sure, this was just for me to get acquainted with the code. Sorting things help me anticipate what the next item in the list will be and allow me to spot missing values, e.g. when there was no `strong` after `strike` I'd directly see that it would be missing.
Reverted in 5d92d00

ljharb replied:
Totally fine to do it in a separate PR, just makes this one harder to review :-)

Primajin replied:
https://github.com/yannickcr/eslint-plugin-react/pull/3182 😉

ljharb commented on lib/rules/no-invalid-html-attribute.js:
```diff
@@ -208,14 +222,16 @@ const COMPONENT_ATTRIBUTE_MAP = new Map();
 COMPONENT_ATTRIBUTE_MAP.set('rel', new Set(['link', 'a', 'area', 'form']));
 
 const messages = {
-  onlyStrings: '“{{attributeName}}” attribute only supports strings.',
-  noEmpty: 'An empty “{{attributeName}}” attribute is meaningless.',
+  emptyIsMeaningless: 'An empty “{{attributeName}}” attribute is meaningless.',
   neverValid: '“{{reportingValue}}” is never a valid “{{attributeName}}” attribute value.',
-  notValidFor: '“{{reportingValue}}” is not a valid “{{attributeName}}” attribute value for <{{elementName}}>.',
-  spaceDelimited: '”{{attributeName}}“ attribute values should be space delimited.',
+  noEmpty: 'An empty “{{attributeName}}” attribute is meaningless.',
   noMethod: 'The ”{{attributeName}}“ attribute cannot be a method.',
+  notAlone: '“{{reportingValue}}” must be directly followed by “{{missingValue}}”.',
+  notPaired: '“{{reportingValue}}” can not be directly followed by “{{secondValue}}” without “{{missingValue}}”.',
+  notValidFor: '“{{reportingValue}}” is not a valid “{{attributeName}}” attribute value for <{{elementName}}>.',
```
also this - let's revert all unrelated formatting/sorting

Primajin replied:
Sure 67b05f7

ljharb reviewed (changes requested):
Let's add some tests that show that `shortcut` by itself is invalid.

Primajin commented:
> Let's add some tests that show that `shortcut` by itself is invalid.

Yes that should be already in:
```js
    {
      code: '<link rel="shortcut"></link>',
      output: '<link rel="shortcut icon"></link>',
      errors: [
        {
          messageId: 'notAlone',
          data: {
            reportingValue: 'shortcut',
            missingValue: 'icon',
          },
        },
      ],
    },
```

Just `shortcut` can not be alone and will be auto-fixed to `shortcut icon`.


Now next I need to look into some spacing that is still added:
![image](https://user-images.githubusercontent.com/1742115/149656988-a036d2d8-0d5d-449b-a747-2d4ef36c4241.png)

Hope to be able to solve this soon 🤞🏻 
When everything is ready I will mark the PR as ... "not draft" 😅 and will write up some description what I did and what my intentions were.

Primajin commented:
OK I see that I can't actually do two things in one go, so even

```js
      code: '<a rel={"batgo noopener"}></a>',
      output: '<a rel={" noopener"}></a>',
```

doesn't turn directly into only `noopener` but it is expected that there is an (invalid) leading space, so then I changed the expectations on my newly added code to also expect a trailing space that I just couldn't manage to get rid of in one go (will be picked up by `spaceDelimited` rule in the next cycle).

I added my intentions kind of line by line so that you can understand the code better for making suggestions.
Happy / Ready for round two 😄 👍🏻

Primajin commented:
Ohh and there was an issue on the white-space part - before we always removed everything that is not exactly one space character - but that means that two spaces between two words will just be removed, instead we should replace all white-space that is not in the beginning or the end to exactly one space character. For security this should also include non-breaking-spaces (e.g. when copied from somewhere else) - so I added another test case that covers that.

ljharb commented:
We should not autofix `shortcut` to `shortcut icon`; that's not a safe autofix because it will change runtime behavior. It should be a warning that requires the user to manually fix it.

ljharb commented on tests/lib/rules/no-invalid-html-attribute.js:
```diff
@@ -373,6 +376,26 @@ ruleTester.run('no-invalid-html-attribute', rule, {
+      ],
+    },
+    {
+      code: '<a rel="noreferrer  noopener"></a>', // this string contains nbsp
```
could we do this using a template literal and `'\xa0'`?

Primajin replied:
It doesn't let me use template literal but it seems to work just fine with single quotes.
![eslint-error](https://user-images.githubusercontent.com/1742115/149669608-f8b2b6d6-dfc1-4a0b-b432-80081824b084.png)
![single-quotes](https://user-images.githubusercontent.com/1742115/149669609-61752c76-7648-44ba-a677-642263accaca.png)

Primajin replied:
Changed in 4de38ef - should work with single quotes as well (test passing locally)

ljharb replied:
oh sure, that works, i forgot this project doesn’t use babel

ljharb commented on tests/lib/rules/no-invalid-html-attribute.js:
```diff
@@ -537,6 +560,74 @@ ruleTester.run('no-invalid-html-attribute', rule, {
       ],
     },
+    {
+      code: '<link rel="shortcut"></link>',
+      output: '<link rel="shortcut icon"></link>',
```
as commented; this isn't a safe autofix.
```suggestion
      code: '<link rel="shortcut"></link>',
```

Primajin replied:
4de38ef

ljharb commented on tests/lib/rules/no-invalid-html-attribute.js:
```diff
@@ -537,6 +560,74 @@ ruleTester.run('no-invalid-html-attribute', rule, {
+      ],
+    },
+    {
+      code: '<link rel="shortcut foo"></link>',
+      output: '<link rel="shortcut icon "></link>',
```
```suggestion
      code: '<link rel="shortcut foo"></link>',
```

Primajin replied:
4de38ef

ljharb commented on tests/lib/rules/no-invalid-html-attribute.js:
```diff
@@ -537,6 +560,74 @@ ruleTester.run('no-invalid-html-attribute', rule, {
+      ],
+    },
+    {
+      code: '<link rel="shortcut  foo"></link>',
+      output: '<link rel="shortcut icon  "></link>',
```
```suggestion
      code: '<link rel="shortcut  foo"></link>',
      output: '<link rel="shortcut foo"></link>',
```
this is a fine autofix, but the test case should still be invalid after it's done.

Primajin replied:
4de38ef

Primajin commented:
Thanks for the feedback! Will work on it! 👍🏻

ljharb commented on tests/lib/rules/no-invalid-html-attribute.js:
```diff
@@ -537,6 +560,73 @@ ruleTester.run('no-invalid-html-attribute', rule, {
+    },
+    {
+      code: '<link rel="shortcut foo"></link>',
+      output: '<link rel="shortcut "></link>',
```
```suggestion
      output: '<link rel="shortcut"></link>',
```

Primajin replied:
Hmm but it actually does that - it's the opposite of

```js
      code: '<a rel={"batgo noopener"}></a>',
      output: '<a rel={" noopener"}></a>',
```

Primajin replied:
```
 1) no-invalid-html-attribute
       invalid
         <link rel="shortcut foo"></link>:

      AssertionError [ERR_ASSERTION]: Output is incorrect.
      + expected - actual

      -<link rel="shortcut "></link>
      +<link rel="shortcut"></link>
```

The test fails when I remove it

Primajin replied:
Might already be covered by

```js
      code: '<a rel="noreferrer noopener foobar"></a>',
      output: '<a rel="noreferrer noopener "></a>',
```
Same behaviour - I could just remove it maybe?

Primajin replied:
Ahh no, we shouldn't remove it - it is the testcase for the `notPaired` branch (in contrast to `notAlone`)

ljharb replied:
I'm basically saying that the autofix output should never have leading or trailing spaces.

Primajin replied:
Yeah that also confused me, but apparently that's "how things used to be" in `master` as well:

https://github.com/yannickcr/eslint-plugin-react/blob/0d8946212a1ba560b12e2e01807033e19dc774f7/tests/lib/rules/no-invalid-html-attribute.js#L355

https://github.com/yannickcr/eslint-plugin-react/blob/0d8946212a1ba560b12e2e01807033e19dc774f7/tests/lib/rules/no-invalid-html-attribute.js#L378

https://github.com/yannickcr/eslint-plugin-react/blob/0d8946212a1ba560b12e2e01807033e19dc774f7/tests/lib/rules/no-invalid-html-attribute.js#L403

https://github.com/yannickcr/eslint-plugin-react/blob/0d8946212a1ba560b12e2e01807033e19dc774f7/tests/lib/rules/no-invalid-html-attribute.js#L416

https://github.com/yannickcr/eslint-plugin-react/blob/0d8946212a1ba560b12e2e01807033e19dc774f7/tests/lib/rules/no-invalid-html-attribute.js#L477

Apparently it only runs one cycle when fixing and we would need to find a way to kind of `while` loop so long until it can't fix any more issues - but that sounds like a major refactor to me 🤔

ljharb replied:
fair point. altho it seems like adding `.trim()` to the attribute contents wouldn't be that difficult?

Primajin replied:
Good idea! But where would I run this, I think I haven't fully grasped when/how fix runs. Would this be part of fix? Or would we just always in the end after all checks run another trim as it's never good/valid to keep it?

ljharb replied:
Adding a single fixer on program exit that trimmed any attribute strings that had been previously autofixed (we’d have to track that manually) would be good. I think that should wait for a followup PR, tho, because it was a pre-existing problem. 

I’ll land this tomorrow, as-is.

ljharb reviewed (approved):
Looks great! Just one nit.

Primajin commented:
Yay 🎉 Thanks for your support - in the end it was a lot of fun, even when it was a bit scary first 🚀

ljharb commented:
Heads up that I did need to land https://github.com/yannickcr/eslint-plugin-react/commit/5f49f51dd237dd4c6758bc0c3e5b300a28770966, due to array destructuring and spread syntax not working on node 4 and 5.

ljharb commented:
I also filed #3180 to track updating the implementation - I think now that we have "pairs", and good test cases, it'd be worth rewriting it from scratch.
